### ISSUE TITLE
npu(moe): 35B-A3B linear-attention conv1d is never applied (host or device) — second missing computation

### ISSUE THREAD

**Issue by bong-water-water-bong** [1bit-MONSTER/1bit-MONSTER#2617]
## Summary

The 35B-A3B MoE layer path has a second missing computation (beyond the routed expert FFN): the linear-attention **causal depthwise conv1d is never applied**, on host or device. This blocks the goal's "correct, reference-matching logits" gate.

## Evidence (re-decoded 2026-09-20, main @ 55ef8f88d)

Re-decoded all 40 `npu-infer/captures/txn-elfs-moe35b/moe_layer_ctx*.txn` with the `decode_txn.cpp` schema (op `0x81` DDR_PATCH; direction from queue-WRITE regs `0x1d204/0x1d20c`=S2MM, `0x1d214/0x1d21c`=MM2S). Every one of the 30 linear-attention layers shows the identical arg3 (norms BO) pattern:

| offset | direction | note |
|---|---|---|
| `0x000000` | **S2MM (write)** | ×2 — the `[0,65536)` `ssm_conv1d` head is written, never read |
| `0x010200` (66048) | MM2S (read) | alpha_proj |
| `0x030200` (197120) | MM2S (read) | beta_proj |
| `0x050200` (328192)+ | MM2S (read) | 34 × 4736-B ssm_out rows |

Confirms #2434 (`RESULTS-moe-conv1d-never-applied-2026-09-16.md`): the conv1d weights are packed into the norms BO head and never consumed. Round 74/75 of `35b-forward-integration.md` mislabeled the `2 × 12288 @0` S2MM write as a read. Host side also has no conv1d (`forward()` contains zero conv code).

## The conv is sandwiched between QKV and SSM — both ELF-internal

The layer ELF computes the fused QKV on-device (act × region-B qkv_proj) and runs the SSM on-device (alpha/beta/ssm_out from the norms BO). The conv1d must run **between** them. Its output contract (confirmed with @agent-d9cc56): `bo_act_ = res + attention + shared_expert` (everything except the routed FFN). So a post-hoc correction is impossible — the no-conv attention is already baked into the residual, and the conv is nonlinear (silu + SSM).

## Fix paths

- **Path A (parity):** the conv1d is a **device sequence**, not a host op — the v0.9.46 lib exports `qwen3_6_moe_npu_sequence::gen_seq_conv1d(seq, 4, 2048, 8192, 4)`. @agent-d9cc56 generated + decoded it (1656 words; DDR_PATCH arg_idx 0 = 8×(base, base+2048) pairs at 4MB steps = the 32MB QKV BO, MM2S read / S2MM write). Fix = insert it between the QKV GEMM (`gen_mm_seq`) and the SSM core on a surfaced QKV BO, since the layer ELF (`_gen_linear_sequence`) fuses QKV+SSM and skips the conv.
- **Path B (correctness oracle):** `~/.dsh/scratch/moe-conv1d/gdn_linear_ref.cpp` — a standalone, compile-checked host reference of the full linear-attention (norm1 → QKV → conv1d → split → alpha/beta → SSM → z-gate → O proj → residual), for diffing the wired output.

## Oracle status (verified)

`~/.dsh/scratch/moe-conv1d/gdn_linear_ref.cpp`:
- Compiles clean (`g++ -O2 -std=c++17 -Wall -Wextra`).
- Smoke-tested with synthetic weights — final residual + stage dumps all finite/nonzero.
- `gdn_attn_cpu` (SSM delta rule) self-checked against a hand-computed case — passes.
- `GDL_REF_DUMP=1` also writes `fqo_pre_conv.bin` [8192], `fqo_post_conv.bin` [8192] (diff against the device `gen_seq_conv1d` stage), and `attn_out_pre_O.bin` [4096].
- O-projection dims cross-checked against the dense path (`OOUT=H=2048`, `OIN=NH*HD=4096`; for GDN `NH*HD == vh*hd = 32*128 = 4096`).

## Exact host port (conv head)

```cpp
// 35B-A3B linear layers: H=2048, qkv_total=8192, vh=32, hd=128, conv_k=4, conv_dim=8192.
// Split q=[0,2048) k=[2048,4096) v=[4096,8192). Per-layer state: conv_state 4×8192 floats
// + delta_state 32×128×128 floats, zero-init, persistent across tokens.

// causal depthwise conv1d on the fused QKV (kernel 4) — npu_engine_universal.cpp:2608-2616
memmove(conv_state, conv_state + dim, (size_t)dim * (k - 1) * sizeof(float));
memcpy(conv_state + (size_t)dim * (k - 1), fqo, (size_t)dim * sizeof(float));
for (int cc = 0; cc < dim; cc++) {
    double s = 0;
    for (int kk = 0; kk < k; kk++)
        s += (double)conv_state[(size_t)kk * dim + cc] * cw[(size_t)kk * dim + cc];
    fqo[cc] = silu((float)s);   // silu: x/(1+expf(-x))
}
// The conv runs over the FULL 8192 fused QKV and only THEN splits q/k/v
// (gdn_attn_step:2618-2630), so do NOT under-conv q/k only.
```

Host weights needed as float (all already touched by `npu_pack_moe_linear5_bo`/`model.c`): `ssm_conv1d` [4,8192], `ssm_alpha_proj`, `ssm_beta_proj`, `ssm_a`, `ssm_dt.bias`, `ssm_norm`, `self_attn.gate_proj`, `ssm_out_proj` [2048,4096], `input_norm`.

See also: `benchmarks/RESULTS-moe35b-conv1d-still-unapplied-2026-09-20.md` (PR #2615).

**Comment by bong-water-water-bong** [comment]
Fixed by #2625 (merged 2026-09-21T03:55:50Z): the host-side Path B (linear_attn_host/std_attn_host) now redoes the full linear attention including conv1d when the layer ELF skips it, instead of leaving it unapplied. Verified on strixhalo hardware: full 40-layer decode reproduces the cached reference (engine_greedy=5470, nonzero=248320, logits corr 0.996806).
